parsec/backend/postgresql/handler.py

```diff
@@ -1,52 +1,103 @@
 # Parsec Cloud (https://parsec.cloud) Copyright (c) AGPLv3 2019 Scille SAS
 
 import trio
+import datetime
 import triopg
-from triopg import UniqueViolationError
+import collections
+from typing import List
+
+from triopg import UniqueViolationError, UndefinedTableError, PostgresError
 from uuid import uuid4
 from functools import wraps
 from structlog import get_logger
 from base64 import b64decode, b64encode
 from importlib_resources import read_text
 
+
 from parsec.event_bus import EventBus
 from parsec.serde import packb, unpackb
 from parsec.utils import start_task
 from parsec.backend.postgresql.tables import STR_TO_REALM_ROLE
-
+from parsec.backend.postgresql import migrations
 
 logger = get_logger()
 
+CREATE_MIGRATION_TABLE_ID = 2
+MigrationResult = collections.namedtuple(
+    "MigrationResult", ["already_applied", "new_apply", "to_apply", "errors"]
+)
 
-async def init_db(url: str) -> None:
+
+async def migrate_db(url: str, migrations: List[str], dry_run: bool) -> MigrationResult:
     """
-    Returns: if the database was already initialized
-    Raises:
-        triopg.exceptions.PostgresError
+    Returns: MigrationResult
     """
     async with triopg.connect(url) as conn:
-        return await _init_db(conn)
-
-
-async def _init_db(conn):
-    if await _is_db_initialized(conn):
-        return True
-
-    async with conn.transaction():
-        query = read_text(__package__, f"init_tables.sql")
-        await conn.execute(query)
-    return False
+        return await _migrate_db(conn, migrations, dry_run)
+
+
+async def _migrate_db(conn, migrations, dry_run):
+
+    errors = []
```

Comment:
Can we have more than one error ?
(I think that if something goes wrong we shouldn't try to go any further)

Reply:
done